### DISCUSSION TITLE
Switching subschemas according to attribute names?

### DISCUSSION THREAD

**Discussion by Markus-Elfring** [json-schema-org/community#202] [Q&A] [answered]
I became curious also about JSON schema composition.
I am looking for another bit of clarification for [the safe application of the key word “oneOf”](https://json-schema.org/understanding-json-schema/reference/combining.html#oneof "Using oneOf").
* Can an attribute name be switched together so that the corresponding subschema can be adjusted?
* Would you like to add any advices for the possibility to distinguish selected data structures based on labels (or property names)?

**Comment by handrews** [comment, marked as answer]
@Markus-Elfring I think you are looking for [`propertyDependencies`](https://github.com/json-schema-org/json-schema-spec/pull/1143) which will be in the next release.

**Reply by handrews** [reply]
Although note that in the meantime you can do something like the following (which will work from draft-06 onwards, or even draft-04 if you replace `const` with a one-element `enum`):

```JSON
"oneOf": [{
  "allOf": [{
    "properties": {"label": {"const": "foo"}}
  }, {
    "$ref": "#/$defs/fooSchema"
  }],
  "allOf": [{
    "properties": {"label": {"const": "bar"}}
  }, {
    "$ref": "#/$defs/barSchema"
  }]
}]
```
which does the same thing as:
```JSON
{
  "propertyDependencies": {
    "label": {
      "foo": {"$ref": "#/$defs/fooSchema"},
      "bar": {"$ref": "#/$defs/barSchema"}
    }
  }
}
```
will do in the next release

**Reply by Markus-Elfring** [reply]
:thought_balloon: How would a similar data structure selection look like if the subschema should be determined for array items depending on labels?

**Reply by handrews** [reply]
@Markus-Elfring can you provide an example schema (minus the selection, obviously- just as close as you can get) and instance?  I'm not quite sure I understand what you mean about the array items and labels.

**Reply by Markus-Elfring** [reply]
>I'm not quite sure I understand what you mean about the array items and labels.

Another example:
```
…
   "label":
…
            {
              "type": "array",
…
              "items":
              { "$ref": "/schemas/Test1" }
            }
…
```


By the way:
Is the text “label” a key word according to the JSON schema specification here?

**Reply by handrews** [reply]
@Markus-Elfring Can you show a complete schema?  It's fine if it's minimal, but I can't tell what's going on here with all of the `...` omissions.  I can't figure out what "label" is doing or even how it relates to the object since it's not the direct value of "labe".  Please show a complete schema (you can leave out stuff compared to your real schema, just make your example complete as it stands) and preferably also the instance you want it to match and how you think the label corresponds to the instance.

**Reply by Markus-Elfring** [reply]
>It's fine if it's minimal, but I can't tell what's going on here with all of the `...` omissions.

:thought_balloon: I would like to clarify design possibilities a bit more also according to the distinction if the desired data type refers to array items or objects altogether.
Such data structures can be denoted with a corresponding identifier which might occasionally be the string “label”.

**Reply by handrews** [reply]
@Markus-Elfring I'm sorry, I've consulted with some other JSON Schema team members and none of us can understand what you want.  You're going to have to show exact schemas and instances to demonstrate this.  I can see that there is a `"type": "array"` and something called `"label"`, but the way you have written this out makes their relationship unclear.

Please write out _complete, syntactically valid_ example schemas and instances if you would like to continue this discussion.

**Reply by Markus-Elfring** [reply]
>I'm sorry, I've consulted with some other JSON Schema team members and none of us can understand what you want. 

I hope that mentioned communication difficulties can be adjusted into more desirable directions.


>I can see that there is a `"type": "array"` and something called `"label"`,

:thought_balloon: Such information is interesting for further clarification, isn't it?


>but the way you have written this out makes their relationship unclear.

I referred to [your suggestion “example schema (minus the selection, obviously- just as close as you can get)”](https://github.com/orgs/json-schema-org/discussions/202#discussioncomment-3232586 "Feedback from 2022-07-26").

**Reply by handrews** [reply]
> I hope that mentioned communication difficulties can be adjusted into more desirable directions.

That would require you to communicate more clearly.  Instead you are just commenting on the confusion.

> 💭 Such information is interesting for further clarification, isn't it?

You've now twice refused to provide the requested information, and comments like this where you seem to be expressing amusement at requests for clarification (which you have also made in other discussions) make it pretty clear that you are not going to provide it at all.

I will be locking this conversation as there does not appear to be a productive way forward here.

**Reply by handrews** [reply]
Edit:  I'll leave this open for 24 hours to give you a chance to respond, but without the requested example or a different and more clear approach from you, I will lock this discussion.

**Reply by Markus-Elfring** [reply]
Will any software users get additional data design ideas from another example like the following?

```JSON
{
  "$schema": "http://json-schema.org/draft/2020-12/schema",
  "$id": "https://example.com/schemas/name_case_distinction-20220805",
  "type": "object",
  "required": ["context"],
  "properties":
  {
    "context":
    {
      "oneOf":
      [
        {
          "allOf":
          [
            {
              "properties":
              {
                "label": {"const": "OB"}
              }
            },
            {
              "$ref": "/schemas/my_customised_object"
            }
          ]
        },
        {
          "allOf":
          [
            {
              "properties":
              {
                "label": {"const": "AI"}
              }
            },
            {
              "type": "array",
              "minItems": 1,
              "items":
              {
                "$ref": "/schemas/my_customised_item"
              }
            }
          ]
        }
      ]
    }
  }
}
```


:crystal_ball:
* How long will it take until the JSON schema feature “propertyDependencies” will be properly supported by selected development tools?
* Will applications of similar case distinctions evolve any further also for more advanced data structure selections?

**Reply by handrews** [reply]
> How long will it take until the JSON schema feature “propertyDependencies” will be properly supported by selected development tools?

First it has to get published in a new draft.  We try to release a significant (as opposed to just wording improvements) new draft roughly annually, usually towards the end of the year.  There's not a set timeline.

After that, it takes a while for tools to implement the new draft - how long varies a great deal depending on the implementation, and you'd have to ask the people maintaining the implementations.

> Will applications of similar case distinctions evolve any further also for more advanced data structure selections?

I have no idea what this is supposed to mean.  JSON Schema keywords are added as people propose specific things and we have a public discussion about it.  Mostly, we now encourage people to build extension vocabularies (since they are now formally supported in 2020-12) and try things out to see how they work.  `propertyDependencies` was an effort to make a better version of OpenAPI's `discriminator` keyword.

**Reply by jdesrosiers** [reply]
You can get the functionality provided by the `propertyDependencies` keyword with the `if`/`then` keywords. It's more complex and verbose, but the functionality is the same. Here's what your example might look like.

```json
{
  "type": "object",
  "properties": {
    "label": { "enum": ["OB", "AI"]
  },

  "allOf": [
    {
      "if": {
        "properties": {
          "label": { "const": "OB" }
        },
        "required": ["label"]
      },
      "then": { "$ref": "/schemas/my_customised_object" }
    },
    {
      "if": {
        "properties": {
          "label": {"const": "AI"}
        },
        "required": ["label"]
      },
      "then": {
        "type": "array",
        "minItems": 1,
        "items": { "$ref": "/schemas/my_customised_item" }
      }
    }
  ]
}
```

You might find the [Applying Subschemas Conditionally](https://json-schema.org/understanding-json-schema/reference/conditionals.html) page of [Understanding JSON Schema](https://json-schema.org/understanding-json-schema/) to be useful.

**Reply by Markus-Elfring** [reply]
>You can get the functionality provided by the `propertyDependencies` keyword with the `if`/`then` keywords.

Thanks for your suggestion of another data design approach.


>It's more complex and verbose,

Will such a view trigger further considerations?


>but the functionality is the same.

:thought_balloon: I suggest to take another look at implementation details for the desired comparison.

* Can it be nicer to move the information “`"type": "object"`” into a corresponding “then” branch?
* Would you like to avoid the duplicate specification of labels?
* Would it be possible to omit the key word “`allOf`” from your example?


I find such case distinctions occasionally useful.
:crystal_ball: I became curious for further collateral evolution (also according to [known software limitations](https://github.com/joelittlejohn/jsonschema2pojo/issues/1184#issuecomment-751297688 "Information for “Conditional subschemas” from 2020-12-25")).

**Reply by handrews** [reply]
@Markus-Elfring all of that is already addressed by `propertyDependencies`.  If you want improvements to "known software limitations", please take that up with the software maintainers.

**Reply by Markus-Elfring** [reply]
>If you want improvements to "known software limitations",

Yes, of course.


>please take that up with the software maintainers.

I am trying also to influence the development situation into desirable directions.

**Reply by handrews** [reply]
> I am trying also to influence the development situation into desirable directions.

I'm not sure why you're trying to do that here.  This is the specification repository.  From the specification point of view, this is solved with `propertyDependencies`, and it will just take time for it to be published an implemented.  Please continue your software-oriented efforts in the software repositories.

**Reply by Markus-Elfring** [reply]
>I'm not sure why you're trying to do that here.

Some adjustment possibilities are in various waiting queues, aren't they?

:thought_balloon: Thus I would like to increase the tolerance and acceptance for desirable changes somehow at several places.

**Reply by handrews** [reply]
There is nothing to do here while waiting for `propertyDependencies`.  If you want tooling to do something better with what currently exists, go talk to the tooling people.  Any spec changes here would not be published before `propertyDependencies`, so there is no reason to do anything here.

**Reply by jdesrosiers** [reply]
> Will such a view trigger further considerations?

Yes, it's easy to end up with bugs if you don't get this pattern exactly right. The link that I provided to the documentation explains all the things you need to watch out for and why in the blue "Note" boxes.

> Can it be nicer to move the information “`"type": "object"`” into a corresponding “then” branch?

This is somewhat a matter of taste. I find that schemas are much more readable, have better error messages, and are often more efficient if you declare your properties first and then complex assertions later. Notice how I use vertical whitespace to separate these parts of the schema. You're example is incomplete, so it's difficult to see the benefits of the pattern in this case. It might even be the case that the pattern doesn't apply this case, but I can't tell either way from the information given.

> Would you like to avoid the duplicate specification of labels?

I would, but with the keywords available now, the best option requires this duplication. `propertyDependencies` was proposed to address this and all the other problems with expressing this kind of constraint. Here you see the duplication disappears with `propertyDependencies` and it's very straightforward to read.

```json
{
  "type": "object",
  "properties": {
    "label": { "enum": ["OB", "AI"] }
  },

  "propertyDependencies": {
    "label": {
      "OB": { "$ref": "/schemas/my_customised_object" },
      "AI": {
        "type": "array",
        "minItems": 1,
        "items": { "$ref": "/schemas/my_customised_item" }
      }
    }
  }
}
```

(I didn't notice before, but your example isn't actually valid. A JSON value can't have property "label" and be an array. It would have to be an object and and array at the same time.)

> Would it be possible to omit the key word “`allOf`” from your example?

No. See https://stackoverflow.com/questions/54694914/multiple-if-then-else-not-validating-for-json-schema/54712494#54712494 for an explanation.

**Reply by Markus-Elfring** [reply]
>There is nothing to do here while waiting for …

:thought_balloon: I hope that development attention will evolve in ways so that application interests can grow considerably.

**Reply by Markus-Elfring** [reply]
>Here you see the duplication disappears with `propertyDependencies` …

How does the string enumeration fit to such information with your example?


>A JSON value can't have property "label" and be an array.

:thought_balloon: I would appreciate further clarification also for this information.


>It would have to be an object and and array at the same time.

* Will it trigger any adjustments for [the documentation](https://github.com/json-schema-org/understanding-json-schema/blob/1a208f8cecd4023db704f4da285fc7148d12de66/source/reference/array.rst#array "Update candidate?")?
* Would any other identifiers (or attribute/property names) reduce understanding difficulties for affected details?


>See https://stackoverflow.com/questions/54694914/multiple-if-then-else-not-validating-for-json-schema/54712494#54712494 for an explanation.

How do you think about the usage of nested constructs “`if`-`then`-`else`” instead of your “`allOf`” example?
JSON code should just be put into the appropriate context and scope, shouldn't it? :thinking:

**Reply by jdesrosiers** [reply]
Most of that I can't answer because I can't figure out what you're asking. For the rest, let's use a different example that is a a little more typical.

```json
{
  "type": "object",
  "properties": {
    "type": { "enum": ["foo", "bar"] },
    "aaa": { "type": "string" },
    "bbb": { "type": "string" }
  },
  "required": ["type"],

  "allOf": [
    { "$ref": "#/$defs/type-foo-requires-aaa" },
    { "$ref": "#/$defs/type-bar-requires-bbb" }
  ],

  "$defs": {
    "type-foo-requires-aaa": {
      "if": {
        "properties": { "type": { "const": "foo" } },
        "required": ["type"]
      },
      "then": { "required": ["aaa"] }
    },
    "type-bar-requires-bbb": {
      "if": {
        "properties": { "type": { "const": "bar" } },
        "required": ["type"]
      },
      "then": { "required": ["bbb"] }
    }
  }
}
```

In this example, I define the full object in the first part, complex assertions in the second part, and and definitions at the end. You can read and understand this schema easily without needing to even look at the complex parts that are abstracted away in the lower parts of the schema.

By defining the complex assertions in definitions, we can give them a descriptive name and use them almost like custom keywords in the schema. You could have written these complex constraints using nested `if`/`then`/`else` instead of two `if`/`then`s, but that would make this schema significantly less readable than the example above.

Any time you see something you that looks like duplication in that schema, it's probably there to improve error messaging when given invalid schemas. In the case of the `"type": { "enum": ["foo", "bar"] }`, this is necessary because the `if`/`then` checks don't constraint "type". Someone could use `"type": "something else"` and the schema wouldn't complain and neither of the complex assertions would assert anything.

The pattern I'm showing isn't always the right fit for every situation. Since your example is incomplete and doesn't make sense to me, I can't tell you if it's the right fit. I can only guess. Hopefully, seeing a full example of the pattern I'm describing makes the other example I gave make a little more sense. Writing a schema is just like writing any other code. There is often more than one way to do things and people's taste and code styles will vary. You aren't bound to use my style if you prefer something else.

**Reply by Markus-Elfring** [reply]
>Any time you see something you that looks like duplication in that schema, it's probably there to improve error messaging when given invalid schemas.

:crystal_ball: How is the software quality evolving for better error reporting approaches also together with JSON schemas?


>Since your example is incomplete

It was constructed for further clarification of questionable implementation details according to another special use case.


>and doesn't make sense to me,

:crystal_ball: How will data design imaginations evolve according to the mentioned case distinctions?
1. Attribute/property names
2. Selection of properties for (JSON) objects or arrays


>I can't tell you if it's the right fit.

:thought_balloon: Was this variant interesting and helpful enough for a constructive discussion?

**Reply by jdesrosiers** [reply]
> Was this variant interesting and helpful enough for a constructive discussion?

I don't know what "variant" you're referring to, but no, nothing you have said has been clear enough for us to have a constructive discussion about. We're doing our best to answer your questions, but we can't understand what you want. I'm sorry, but I have to be done with this conversation for now because it hasn't been constructive and the way it's been going, I don't see that improving.

**Reply by Markus-Elfring** [reply]
>…, I don't see that improving.

:thought_balloon: Do you find another clarification approach worth for further development considerations?

[Checking support for the key words “oneOf” and “allOf”](https://github.com/eclipse-vertx/vertx-json-schema/issues/72 "Clarification request from 2022-11-14")

**Reply by Markus-Elfring** [reply]
:thought_balloon: Would you like to add any comments to another clarification approach?

[Improve data type determination for the key words “oneOf” and “allOf”](https://github.com/ajv-validator/ajv-cli/issues/224 "Clarification request from 2022-11-15")
